### PR TITLE
Exception for the intersectionWith method

### PR DESCRIPTION
**Pull request by Richiedu38**
To avoid error messages during compilation due to the absence of the intersectionWith method for some shapes, I added an exception returning 'Intersection line capsule isn't supported'.

### CODE REVIEW

File: src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/Face3DReadOnly.java [hunk 1]
```diff
@@ -688,6 +690,35 @@ default boolean orthogonalProjection(Point3DReadOnly pointToProject, Point3DBasi
...
+      if (intersectionWithPlane) {
+         
+         boolean insideTheFace = isPointInside(intersectionToPack, EuclidGeometryTools.ONE_MILLIONTH);
+         
+         if (insideTheFace)
+            return true;
+   
+      }
+      
+      return false;
```
Comment (SylvainBertrand) [inline src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/Face3DReadOnly.java:717]: To be consistent with the other intersection algos, let's call `intersectionToPack.setToNaN()` when it's not intersecting.

File: src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/ConvexPolytope3DReadOnly.java [hunk 2]
```diff
@@ -154,6 +156,52 @@ default void getBoundingBox(BoundingBox3DBasics boundingBoxToPack)
...
+   
+   @Override
+
+   default int intersectionWith(Point3DReadOnly pointOnLine,
+                                Vector3DReadOnly lineDirection,
+                                Point3DBasics firstIntersectionToPack,
+                                Point3DBasics secondIntersectionToPack)
+   
+   {
+
```
Comment (SylvainBertrand) [inline src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/ConvexPolytope3DReadOnly.java:170]: let's add here:
```
if (firstIntersectionToPack == null) firstIntersectionToPack = new Point3D();
if (secondIntersectionToPack == null) secondIntersectionToPack = new Point3D();
```

File: src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/ConvexPolytope3DReadOnly.java [hunk 3]
```diff
@@ -154,6 +156,52 @@ default void getBoundingBox(BoundingBox3DBasics boundingBoxToPack)
...
+      if (firstIntersectionToPack != null)
+         firstIntersectionToPack.setToNaN();
+      if (secondIntersectionToPack != null)
+         secondIntersectionToPack.setToNaN();
+      
+      int numberOfIntersections = 0;
+      
+      //For each face, we check the possibility of intersection with the line
+      for (int i=0; i< getNumberOfFaces(); i++) {
+         //If there isn't already a intersection found
```
Comment (SylvainBertrand) [inline src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/ConvexPolytope3DReadOnly.java:180]: The conditions in this loop aren't working, also you've got a bug due to a bad copy-paste. I'll let you figure out how to fix this loop.

File: src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/ConvexPolytope3DReadOnly.java [hunk 4]
```diff
@@ -154,6 +156,52 @@ default void getBoundingBox(BoundingBox3DBasics boundingBoxToPack)
...
+            if (intersectionWithFace) {
+               numberOfIntersections++;   
+            }
+         }
+         
+         if (firstIntersectionToPack != null && secondIntersectionToPack != null)
+            break;
+      }
+      
+      return numberOfIntersections;
```
Comment (SylvainBertrand) [inline src/shape/java/us/ihmc/euclid/shape/convexPolytope/interfaces/ConvexPolytope3DReadOnly.java:199]: Before returning, we need to check the ordering of the 2 intersections to be consistent with the other intersection algorithms

**Review by SylvainBertrand — CHANGES_REQUESTED** [review]
I only commented on the polytope stuff

File: src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java [hunk 5]
```diff
@@ -1906,6 +1946,215 @@ else if (convexPolytope3D.getNumberOfVertices() == 2)
...
+
+         double alpha = random.nextDouble();
+         double beta = random.nextDouble() * (1 - alpha);
+         double gamma = 1 - alpha - beta;
+
+         double x = alpha * A.getX() + beta * B.getX() + gamma * C.getX();
+         double y = alpha * A.getY() + beta * B.getY() + gamma * C.getY();
+         double z = alpha * A.getZ() + beta * B.getZ() + gamma * C.getZ();
+
+         Point3D pointOnFace = new Point3D(x, y, z);
```
Comment (SylvainBertrand) [inline src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java:2061]: I'd suggest wrapping the above code for generating a random point in a face into a static method here.
Eclipse can help you do that, select the code, then right click, refactor > extract method

File: src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java [hunk 6]
```diff
@@ -1906,6 +1946,215 @@ else if (convexPolytope3D.getNumberOfVertices() == 2)
...
+
+         Point3D actualFirstInstersection = new Point3D();
+         Point3D actualSecondInstersection = new Point3D();
+
+         assertEquals(0, convexPolytope3D.intersectionWith(pointOutside, lineDirection, null, null));
+         assertEquals(0, convexPolytope3D.intersectionWith(pointOutside, lineDirection, actualFirstInstersection, null));
+         assertEquals(0, convexPolytope3D.intersectionWith(pointOutside, lineDirection, null, actualSecondInstersection));
+         actualFirstInstersection.setToZero();
+         actualSecondInstersection.setToZero();
+         assertEquals(0, convexPolytope3D.intersectionWith(pointOutside, lineDirection, actualFirstInstersection, actualSecondInstersection));
```
Comment (SylvainBertrand) [inline src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java:2081]: I think we can find more constructions to cover more scenarios with no intersection

File: src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java [hunk 7]
```diff
@@ -1906,6 +1946,215 @@ else if (convexPolytope3D.getNumberOfVertices() == 2)
...
+
+         Point3D actualFirstInstersection = new Point3D();
+         Point3D actualSecondInstersection = new Point3D();
+
+         assertEquals(2, convexPolytope3D.intersectionWith(pointInside, lineDirection, null, null));
+         assertEquals(2, convexPolytope3D.intersectionWith(pointInside, lineDirection, actualFirstInstersection, null));
+         assertEquals(2, convexPolytope3D.intersectionWith(pointInside, lineDirection, null, actualSecondInstersection));
+         actualFirstInstersection.setToZero();
+         actualSecondInstersection.setToZero();
+         assertEquals(2, convexPolytope3D.intersectionWith(pointInside, lineDirection, actualFirstInstersection, actualSecondInstersection));
```
Comment (SylvainBertrand) [inline src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java:2152]: Here you've got the intersections that are known, you should easily be able to check the ones that are computed

File: src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java [hunk 8]
```diff
@@ -1906,6 +1946,215 @@ else if (convexPolytope3D.getNumberOfVertices() == 2)
...
+
+      }
+
```
Comment (SylvainBertrand) [inline src/test/java/us/ihmc/euclid/shape/convexPolytope/ConvexPolytope3DTest.java:2155]: Missing test for 1 intersection: through a vertex and through an edge.
Missing annoying edge-case: the line is in-plane of a face, there should be 2 intersections still

File: src/geometry/java/us/ihmc/euclid/geometry/tools/EuclidGeometryTools.java [hunk 9]
```diff
@@ -4040,6 +4043,73 @@ public static boolean intersectionBetweenLine2DAndLineSegment2D(Point2DReadOnly
...
+    */
+   public static int intersectionBetweenLine3DAndPointShape3DImpl(double pointX,
+                                                                 double pointY,
+                                                                 double pointZ,
+                                                                 double lineStartX,
+                                                                 double lineStartY,
+                                                                 double lineStartZ,
+                                                                 double lineDirectionX,
+                                                                 double lineDirectionY,
+                                                                 double lineDirectionZ,
```
Comment (SylvainBertrand) [inline src/geometry/java/us/ihmc/euclid/geometry/tools/EuclidGeometryTools.java:4087]: This is only testing for the distance between line and point, not sure that's worth implementing a new method for it, we could simply rely on the distance method directly and implement the intersection in `PointShape3DReadOnly` directly.

**Review by SylvainBertrand — CHANGES_REQUESTED** [review]
